test/screenshot/lib/user-agent.js

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable key-spacing, no-unused-vars */
+
+const CbtApi = require('./cbt-api');
+const cbtApi = new CbtApi();
+
+const CBT_FILTERS = {
+  formFactor: {
+    any:     () => (device) => true,
+    desktop: () => (device) => device.device === 'desktop',
+    mobile:  () => (device) => device.device === 'mobile' && device.device_type !== 'tablet',
+  },
+
+  operatingSystemName: {
+    any:     () => (device) => true,
+    android: () => (device) => device.type === 'Android',
+    ios:     () => (device) => device.type === 'iPhone',
+    mac:     () => (device) => device.type === 'Mac',
+    windows: () => (device) => device.type === 'Windows',
+  },
+
+  browserName: {
+    any:     () => (browser) => true,
+    chrome:  () => (browser) => browser.type === 'Chrome' || browser.type === 'Chrome Mobile',
+    edge:    () => (browser) => browser.type === 'Microsoft Edge',
+    firefox: () => (browser) => browser.type === 'Firefox',
+    ie:      () => (browser) => browser.type === 'Internet Explorer',
+    opera:   () => (browser) => browser.type === 'Opera',
+    safari:  () => (browser) => browser.type === 'Safari' || browser.type === 'Mobile Safari',
+  },
+
+  browserVersion: {
+    any:      (desiredVersion) => (actualVersion) => true,
+    exactly:  (desiredVersion) => (actualVersion) => compareVersions(actualVersion, desiredVersion) === 0,
+    latest:   (n = 1) => (actualVersion, actualVersionIndex, actualVersionListSortedDesc) => {
+      // Array is sorted by `filterBrowsersByVersion()`
+      return actualVersionListSortedDesc
+        .slice(0, n)
+        .includes(actualVersion)
+      ;
+    },
+    previous: (n = 1) => (actualVersion, actualVersionIndex, actualVersionListSortedDesc) => {
+      // Array is sorted by `filterBrowsersByVersion()`
+      return actualVersionListSortedDesc
+        .slice(1, 1 + n)
+        .includes(actualVersion)
+      ;
+    },
+  },
+};
+
+let userAgentConfigsPromise;
+
+module.exports = {
+  fetchConfigs,
+  fetchConfigFromApiName,
+};
+
+async function fetchConfigs() {
```

Comment:
Does this map CBT-parsable user agents to our own syntax for user agents?

Reply:
Yes, exactly!

Most of the `device` and `browser` properties of a "config" object come straight from CBT. I just shuffle things around a bit to make it easier to work with.

I realize it might be a bit confusing, but I couldn't think of a better way to do it and keep it simple.

Do you have any suggestions?

Reply:
K, I:

- Renamed `user-agent.js` to `cbt-user-agent.js`
- Renamed `fetchConfigs` to `fetchBrowsersToRun`
- Created `@typedef` annotations for all the anonymous objects:
    - `CbtUserAgent`
    - `CbtDevice`
    - `CbtBrowser`
    - `CbtDeviceResolution`

Once #2687 is merged, I can move the `@typedef`s to `types.js`. For now, I just put them at the bottom of `cbt-user-agent.js`.

Does that seem reasonable?